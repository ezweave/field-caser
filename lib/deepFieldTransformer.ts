import { isArray, isObject, map, partialRight, reduce } from 'lodash'
/**
 * This version of lodash/fp seems to muck up the arguments for reduce.  It does
 * not pass the key.
 *
 * @param identitee
 * @returns {*}
 */

const reduceFP = (func, collector) => partialRight(reduce, func, collector)

export type StringTransformer = (string) => string

export type DeepFieldTransformer = <T>(stringTransformer: StringTransformer) => (source: unknown) => T

export const deepFieldTransformer: DeepFieldTransformer = (stringTransformer) =>
  reduceFP(
    (result, value, key) => ({
      ...result,
      [stringTransformer(key)]: isArray(value)
        ? map(value, (entry) => deepFieldTransformer(stringTransformer)(entry))
        : isObject(value)
        ? deepFieldTransformer(stringTransformer)(value)
        : value,
    }),
    {}
  )
