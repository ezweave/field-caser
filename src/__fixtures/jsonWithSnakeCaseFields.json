{
	"first_name": "Joe",
	"last_name": "Consumer",
	"details": {
		"date_of_birth": "1/1/1970",
		"cars": [{
				"brand": "Pontiac",
				"model": "Aztek",
				"year_of_manufacture": "2004"
			},{
				"brand": "Saturn",
				"model": "Vue",
				"year_of_manufacture": "2004"
		}]
	}
}
