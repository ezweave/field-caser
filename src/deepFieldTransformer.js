import { isArray, isObject, map, partialRight, reduce } from 'lodash';
/**
 * This version of lodash/fp seems to muck up the arguments for reduce.  It does
 * not pass the key.
 *
 * @param identitee
 * @returns {*}
 */

const reduceFP = (func, collector) => partialRight(reduce, func, collector);

export const deepFieldTransformer = fieldConverter => reduceFP((result, value, key) => ({ ...result,
  [fieldConverter(key)]: isObject(value) ? deepFieldTransformer(fieldConverter)(value) : isArray(value) ? map(value, deepFieldTransformer(fieldConverter)) : value
}), {});