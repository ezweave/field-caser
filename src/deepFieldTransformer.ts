import { isArray, isObject, map, partialRight, reduce } from 'lodash'
/**
 * This version of lodash/fp seems to muck up the arguments for reduce.  It does
 * not pass the key.
 *
 * @param identitee
 * @returns {*}
 */

const reduceFP = (func, collector) => partialRight(reduce, func, collector)

export interface StringTransformer {
  (string): string
}

export interface DeepFieldTransformer<T, O> {
  (stringTransformer: StringTransformer): (T) => O
}

export const deepFieldTransformer: DeepFieldTransformer<{}, {}> = stringTransformer => reduceFP((result, value, key) => ({ ...result,
  [stringTransformer(key)]: isObject(value) ? deepFieldTransformer(stringTransformer)(value) : isArray(value) ? map(value, deepFieldTransformer(stringTransformer)) : value
}), {})