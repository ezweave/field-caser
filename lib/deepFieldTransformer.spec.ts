import { toUpper } from 'lodash'
import { deepFieldTransformer } from './deepFieldTransformer'
import * as jsonWithSnakeCaseFields from './__fixtures/jsonWithSnakeCaseFields.json'

describe('deepFieldTransformer', () => {
  it('converts nested field names to using a converter', () => {
    expect(deepFieldTransformer(toUpper)(jsonWithSnakeCaseFields)).toMatchSnapshot()
  })
})