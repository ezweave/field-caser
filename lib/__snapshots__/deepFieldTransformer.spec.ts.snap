// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deepFieldTransformer converts nested field names to using a converter 1`] = `
Object {
  "DEFAULT": Object {
    "DETAILS": Object {
      "CARS": Object {
        "0": Object {
          "BRAND": "Pontiac",
          "MODEL": "Aztek",
          "YEAR_OF_MANUFACTURE": "2004",
        },
        "1": Object {
          "BRAND": "Saturn",
          "MODEL": "Vue",
          "YEAR_OF_MANUFACTURE": "2004",
        },
      },
      "DATE_OF_BIRTH": "1/1/1970",
    },
    "FIRST_NAME": "Joe",
    "LAST_NAME": "Consumer",
  },
  "DETAILS": Object {
    "CARS": Object {
      "0": Object {
        "BRAND": "Pontiac",
        "MODEL": "Aztek",
        "YEAR_OF_MANUFACTURE": "2004",
      },
      "1": Object {
        "BRAND": "Saturn",
        "MODEL": "Vue",
        "YEAR_OF_MANUFACTURE": "2004",
      },
    },
    "DATE_OF_BIRTH": "1/1/1970",
  },
  "FIRST_NAME": "Joe",
  "LAST_NAME": "Consumer",
}
`;
